src/js_native_api_v8.cc

```diff
@@ -681,6 +681,155 @@ inline napi_status Wrap(napi_env env,
+                                   napi_value dest,
+                                   size_t static_property_count,
+                                   size_t property_count,
+                                   const napi_property_descriptor* properties) {
+  if (static_property_count > 0) {
+    std::vector<napi_property_descriptor> static_descriptors;
+    static_descriptors.reserve(static_property_count);
+
+    for (size_t i = 0; i < property_count; i++) {
+      const napi_property_descriptor* p = properties + i;
+      if ((p->attributes & napi_static) != 0) {
+        static_descriptors.push_back(*p);
+      }
+    }
+
+    STATUS_CALL(napi_define_properties(env,
+                                       dest,
+                                       static_descriptors.size(),
+                                       static_descriptors.data()));
+  }
+  return napi_clear_last_error(env);
+}
+
+class JSClassGen {
+ public:
+  JSClassGen(const char* utf8name, size_t length) {
+    if (length == NAPI_AUTO_LENGTH)
+      classname = std::string(utf8name);
+    else
+      classname = std::string(utf8name, length);
+    args.resize(3);
+  }
+
+  inline void
+  PostProcess(const std::string& key, const napi_property_descriptor* prop) {
+    if (prop->attributes == napi_default && prop->value == nullptr) return;
+
+    post_process += "    " + key + ": {\n";
+    if (prop->value != nullptr)
+      post_process += "      value: " + PushArg(prop->value) + ",\n";
+    if (prop->attributes & napi_writable)
+      post_process += "      writable: true,\n";
+    if (prop->attributes & napi_configurable)
+      post_process += "      configurable: true,\n";
+    if (prop->attributes & napi_enumerable)
+      post_process += "      enumerable: true,\n";
+    post_process += "    },\n";
+  }
+
+  inline std::string PushArg(napi_value val = nullptr) {
+    std::string propname = "prop" + std::to_string(param_idx++);
+    formal_args += ", " + propname;
+    if (val != nullptr) args.push_back(val);
+    return propname;
+  }
+
+  inline napi_status PushFunction(napi_env env,
+                                  const napi_property_descriptor* desc,
+                                  napi_callback method,
+                                  const std::string& key) {
+    if (method == nullptr) return napi_ok;
+
+    napi_value fn;
+    STATUS_CALL(napi_create_function(env,
+                             desc->utf8name,
+                             (desc->utf8name == nullptr ? 0 : NAPI_AUTO_LENGTH),
+                             method,
+                             desc->data,
+                             &fn));
+    args.push_back(fn);
+
+    class_body += "    " +
+        ((method == desc->method) ? key + "(...x)" :
+        (method == desc->getter) ? "get " + key + "()" :
+        "set " + key + "(x)") + " {\n"
+        "      if (!(this instanceof " + classname + "))\n"
+        "        throw new Error('Illegal invocation');\n"
```

Comment:
Why are we adding this brand check? JS classes don’t do it, and for C++ classes, it’s not necessarily the right thing to do in the presence of prototype manipulation

Reply:
Classes created using `napi_define_class` have such a brand check built-in to ensure that internal field counts line up. The engine does this. Check the `assert.throws()` I added to test_constructor. It also throws for a class created using `napi_define_class`, which I have not modified (other than to factor out the code that adds static members for reuse with `napi_define_subclass`.

Reply:
Yeah, the point is that this does not actually ensure that the internal fields are present as expected.

Reply:
@addaleax in that case I think I need to back it up with type tags and throw the exception if the tag is absent. This also means that I have to likely subclass `v8impl::CallbackWrapper` to do the check before calling into the add-on.

Reply:
@addaleax ... or, rather, replace the `instanceof` check in the JS snippet with a type-tag check on the native side.

Reply:
@addleax ... which also means that, for efficiency's sake, it might be best located in core 😉🙂

Reply:
I think if we’re talking efficiency then generating a new script per subclass is already throwing that out of the window ;)

Reply:
🙂

Reply:
@addaleax actually, unless one generates subclasses in a tight loop, only the startup performance is hit by the need to compile JS code. I guess I was talking about the price for a call.

```
JS -> v8impl::CallbackWrapper ----> (napi_env env, napi_callback_info info) -> (napi_env env, napi_callback_info info)
                                    type tag check happens here
```

as would be an external implementation, is more expensive than

```
JS -> v8impl::CallbackWrapper ----> (napi_env env, napi_callback_info info)
      type tag check happens here
```

which is how it would be done internally.